.github/workflows/changelogs/changelogs.py

```diff
@@ -0,0 +1,639 @@
+        for kind, regex in self.regex.trackers.items():
+            trackers_of_kind = []
+            for match in re.finditer(regex, text):
+                # Match groups are defined by contract in the tracker file
+                # Group 1 should be the numeric ID of the tracker
+                trackers_of_kind.append(match.group(0, 1))
+
+            # Gather the unique trackers
+            trackers[kind] = list(set(trackers_of_kind))
+
+        return trackers
+
+    def get_pr_trackers(self, git_repo: str, pr_number: int) -> dict[str, list[tuple[str, str]]]:
+        """Get all the trackers mentioned in a PR
+
+        The trackers are extracted from the PR title and the commit messages.
+        """
+
+        assert pr_number
+        assert git_repo
+
+        logging.info(f"Requesting information for PR#{pr_number} at '{git_repo}'")
+        stream = os.popen(f'gh pr view -R {git_repo} {pr_number} --json title,commits -q ".title, .commits[].messageHeadline, .commits[].messageBody | select(length > 0)"')
+        commits = stream.read()
+        if stream.close():
+            raise Exception("An error occurred when getting the PR information from the GitHub API.")
+        return self.extract_trackers(commits)
+
+    def validate_chlog_entry(self, entry: type[Entry]) -> list[type[Issue]]:
+        """Validate a single changelog entry"""
+
+        issues = []
+        # Test capitalization
+        if re.search(self.regex.WRONG_CAP_START, entry.entry) or re.search(self.regex.WRONG_CAP_AFTER, entry.entry):
+            issues.append(Issue(IssueType.WRONG_CAP, entry.file, entry.line, entry.end_line))
+        # Test spacing
+        if re.search(self.regex.WRONG_SPACING, entry.entry):
+            issues.append(Issue(IssueType.WRONG_SPACING, entry.file, entry.line, entry.end_line))
+
+        return issues
+
+    def get_entry_obj(self, buffer: list[str], file: str, line_no: int) -> type[Entry]:
+        """Create an Entry object from a buffer of entry lines
+
+        The elements in the 'buffer' list are separate lines of a single entry.
+        """
+
+        # Strip the '- ' characters in the beginning of the first line
+        msg = ''.join(buffer)[2:]
+        trackers = self.extract_trackers(msg)
+        return Entry(msg, file, line_no - len(buffer), line_no - 1 if len(buffer) > 1 else None, trackers)
+
+    def validate_chlog_file(self, file: str) -> tuple[list[type[Issue]], list[type[Entry]]]:
+        """Validate a single changelog file"""
+
+        logging.debug(f"Validating changelog file: {file}")
+        file_path = os.path.join(self.spacewalk_root, file)
+
+        if os.path.getsize(file_path) == 0:
+            return ([Issue(IssueType.EMPTY_CHLOG, file)], [])
+
+        f = open(file_path, "r")
+        issues = []
+        entries = []
+        entry_buf: list[str] = [] # List to buffer the lines in a single changelog entry
+        line_no = 0
+
+        for line in f:
+            line_no += 1
```

Comment:
You don't have to count manually, `enumerate(iterable)` does it for you :)
```suggestion
        for line_no, line in enumerate(f):
```

Reply:
neat :)